processor/queuedprocessor/queued_processor.go

```diff
@@ -16,76 +16,209 @@ package queuedprocessor
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"sync"
 	"time"
 
 	"github.com/jaegertracing/jaeger/pkg/queue"
 	"go.opencensus.io/stats"
-	"go.opencensus.io/stats/view"
 	"go.opencensus.io/tag"
 	"go.uber.org/zap"
 
 	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/component/componenterror"
 	"go.opentelemetry.io/collector/consumer"
 	"go.opentelemetry.io/collector/consumer/consumererror"
 	"go.opentelemetry.io/collector/consumer/pdata"
-	"go.opentelemetry.io/collector/internal/collector/telemetry"
+	"go.opentelemetry.io/collector/consumer/pdatautil"
 	"go.opentelemetry.io/collector/obsreport"
 	"go.opentelemetry.io/collector/processor"
 )
 
-type queuedSpanProcessor struct {
+type queuedProcessor struct {
 	name                     string
 	queue                    *queue.BoundedQueue
 	logger                   *zap.Logger
-	sender                   consumer.TraceConsumer
+	traceNext                consumer.TraceConsumer
+	metricNext               consumer.MetricsConsumer
 	numWorkers               int
 	retryOnProcessingFailure bool
 	backoffDelay             time.Duration
-	statsTags                []tag.Mutator
 	stopCh                   chan struct{}
 	stopOnce                 sync.Once
 }
 
-var _ consumer.TraceConsumer = (*queuedSpanProcessor)(nil)
+var _ consumer.TraceConsumer = (*queuedProcessor)(nil)
+var errorRefused = errors.New("failed to add to the queue")
+
+type queueItem interface {
+	context() context.Context
+	queuedTime() time.Time
+	export(sp *queuedProcessor) error
+	onAccepted()
+	onPartialError(partialErr consumererror.PartialError)
+	onRefused(logger *zap.Logger, err error)
+	onDropped(logger *zap.Logger, err error)
+}
+
+type baseQueueItem struct {
+	ctx context.Context
+	qt  time.Time
+}
+
+func (item *baseQueueItem) context() context.Context {
+	return item.ctx
+}
+
+func (item *baseQueueItem) queuedTime() time.Time {
+	return item.qt
+}
 
-type queueItem struct {
-	queuedTime     time.Time
+type traceQueueItem struct {
+	baseQueueItem
 	td             pdata.Traces
 	spanCountStats *processor.SpanCountStats
-	ctx            context.Context
 }
 
-func newQueuedSpanProcessor(
+func newTraceQueueItem(ctx context.Context, td pdata.Traces, instanceName string) queueItem {
+	// Ensure always have the TagProcessorNameKey set in the context.
+	// TODO: Remove this after #1312 is merged and verify.
+	ctx, _ = tag.New(ctx, tag.Upsert(processor.TagProcessorNameKey, instanceName))
```

Comment:
do not forget to check this before merge since #1312 is already in

Reply:
Done.